resources/js/components/news-announcements.tsx

```diff
@@ -0,0 +1,199 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the GNU Affero General Public License v3.0.
+// See the LICENCE file in the repository root for full licence text.
+
+import Img2x from 'components/img2x';
+import NewsAnnouncementJson from 'interfaces/news-announcement-json';
+import { range } from 'lodash';
+import { action, makeObservable, observable } from 'mobx';
+import { observer } from 'mobx-react';
+import * as React from 'react';
+import { classWithModifiers } from 'utils/css';
+
+function modulo(dividend: number, divisor: number): number {
+  return ((dividend % divisor) + divisor) % divisor;
+}
+
+const autoRotateIntervalMs = 6000;
+const bn = 'news-announcements';
+const itemBn = 'news-announcement';
+
+interface Props {
+  announcements: NewsAnnouncementJson[];
+}
+
+@observer
+export default class NewsAnnouncements extends React.Component<Props> {
+  private autoRotateTimerId?: number;
+  @observable private index = 0;
+  @observable private maxIndex = this.length - 1;
+  @observable private minIndex = 0;
+  @observable private transition = true;
+
+  private get length() {
+    return this.props.announcements.length;
+  }
+
+  constructor(props: Props) {
+    super(props);
+
+    makeObservable(this);
+  }
+
+  componentDidMount() {
+    this.setAutoRotateTimer();
+  }
+
+  componentWillUnmount() {
+    this.clearAutoRotateTimer();
+  }
+
+  render() {
+    if (this.length === 0) {
+      return null;
+    }
+
+    if (this.length === 1) {
+      return (
+        <div className={bn}>
+          <div className={`${bn}__container`}>
+            {this.renderAnnouncement(this.props.announcements[0])}
+          </div>
+        </div>
+      );
+    }
+
+    return (
+      <div
+        className={bn}
+        onMouseEnter={this.clearAutoRotateTimer}
+        onMouseLeave={this.setAutoRotateTimer}
+      >
+        <div
+          className={classWithModifiers(`${bn}__container`, { transition: this.transition })}
+          onTransitionEnd={this.handleTransitionEnd}
+          style={{ '--index': this.index } as React.CSSProperties}
+        >
+          {/*
+            Render the announcements, including clones before and after to help
+            create the illusion of an infinitely scrolling container
+          */}
```

Comment:
where's the clone? the range below only generates exactly the length of announcements.

From what I can tell the illusion currently happens because the clone is generated exactly before the animation starts?